frescobaldi_app/vcs/gitrepoo.py

```diff
@@ -0,0 +1,416 @@
+    # 5.  parse git diff output
+    # 6.  current_branch() 
+    # 7.  branches()
+    # 8.  checkout()
+    # 9.  has_branch()
+    # 10. has_remote() 
+    # 11. has_remote_branch()
+    # 12. remotes()
+    # 13. tracked_remote()
+    # 14. tracked_remote_label()
+    # 15. set_working_directory()
+    # 16. whether the file is an untracked file
+    # 17. whether the file is an ignored file
+    # 18. whether git has been installed 
+    # 19. use which git
+    # 19. get git version
+    ############################################################################ 
+
+    def __init__(self, current_view = None):
+        # update signal emits signal when git content updates
+        self._update_signal = pyqtSignal(bool)
+        # git command instance
+        self._git = gitjob.Git()
+        # Identifier of comparison object
+        # 'Index' for Index
+        # 'HEAD' for HEAD 
+        # commit-hash for specific commit
+        self._compare_against = 'HEAD'
+        # root path of current working directory
+        self._root_path = None
+        # current file/directory's relative path
+        self._relative_path = None
+        self._temp_committed_file = None 
+        self._temp_Index_file = None
+        self._temp_working_file = None  
+        # current file's view instance
+        # None: If the instance is not operating on a file
+        self._current_view = current_view
+    
+    def __del__(self):
+        """Caller is responsible for clean up the temp files"""
+        if self._temp_committed_file:
+            os.unlink(self._temp_committed_file)
+        if self._temp_Index_file:
+            os.unlink(self._temp_Index_file)
+        if self._temp_working_file:
+            os.unlink(self._temp_working_file)
+
+    def set_working_directory(self, path):
+        """
+        Set the _root_path and _relative_path
+        Return True if giving path is git initialized.  
+        """
+        self._root_path, self._relative_path = self._get_work_path(path)
+        return self._root_path != None
+       
+
+    def _is_git_path(self, path):
+        """
+        Return True if 'path' is a valid git working tree.
+        """
+        return path and os.path.exists(os.path.join(path, '.git'))
+    
+
+    def _get_work_path(self, ori_path):
+        """
+        If the giving ori_path is git initialized. 
+        Split the ori_path into the working tree part and relative path part.
+        This function handles both file pathes and directory pathes.
+
+        Note:
+            This is a local alternative to calling the git command:
+
+                git rev-parse --show-toplevel
+
+        Arguments:
+            ori_path (string): a absolute path.
+
+        Returns:
+            (root path of the working proj, relative path)
+            (None, None) for non-git-initilized file_path
+        """
+        if ori_path and os.path.exists(ori_path):
+            if os.path.isdir(ori_path):
```

Comment:
This doesn't really work. `os.path.isdir` doesn't care about the presence or absence of a trailing (back)slash while your logic seems to rely on that difference.

If I understand better what the caller of the function actually expects I can make a better comment.